src/Microsoft.ML.LightGBM/LightGbmTrainerBase.cs

```diff
@@ -44,8 +46,8 @@ private sealed class CategoricalMetaData
         /// the code is culture agnostic. When retrieving key value from this dictionary as string
         /// please convert to string invariant by string.Format(CultureInfo.InvariantCulture, "{0}", Option[key]).
         /// </summary>
-        private protected readonly Dictionary<string, object> Options;
-        private protected readonly IParallel ParallelTraining;
+        private protected Dictionary<string, object> Options;
+        private protected IParallel ParallelTraining;
```

Comment:
It would be nice to keep them readonly. I see the problem why you made them non-readonly.
May be call the second constructor from the first one and create `Args` in static method for passing into the second one.
This way you can initialize them in the second constructor only.

Reply:
Llike the idea, and looked into, but I can't do it because the two constructor call different base class constructors, and can't change that.. 

---
In reply to: [219271083](https://github.com/dotnet/machinelearning/pull/962#discussion_r219271083) [](ancestors = 219271083)